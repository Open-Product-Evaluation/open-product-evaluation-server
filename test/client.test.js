const users = require('../seeds/data/user/user')
const clients = require('../seeds/data/client/client')
const domains = require('../seeds/data/domain/domain')
const { seedDatabase } = require('mongo-seeding')
const config = require('../config')
const request = require('./requesthelper')
const { getSeedID } = require('./helpers')

/* Functions for Querys */

function createClientQuery(name) {
  return {
    query: `mutation {
        createClient(data: {name: "TestClient"}) {
          client {
            name
            domain {
              id
            }
          }
          token
        }
      }`,
  }
}

function updateClientQuery(clientID, clientName, domain, owners) {
  return {
    query: `mutation {
      updateClient(clientID:"${clientID}", data:{name:"${clientName}",domain:"${domain}", owners: ${JSON.stringify(owners)} }){
        client{
          id
          name
          domain{
            id
          }
          owners{
            id
          }
        }
      }
    }`,
  }
}

function deleteClientQuery(clientID) {
  return {
    query: `mutation {
      deleteClient(clientID:"${clientID}"){
        success
      }
    }`,
  }
}

function clientsQuery() {
  return {
    query: `{
      clients {
        id
        name
        domain {
          id
        }
        owners {
          id
        }
      }
    }`,
  }
}

function clientQuery(clientID) {
  return {
    query: `{
      client(clientID: "${clientID}") {
        id
        name
        domain {
          id
        }
        owners{
          id
        }
      }
    }`,
  }
}

/* Tests */

describe('Client', () => {
  describe('Anonym', async () => {
    beforeAll(() => seedDatabase(config.seeder))
    it('should create a Client [Mutation]', async () => {
      const query = createClientQuery('TestDevie')
      const res = await request.anon(query)
      const { data, errors } = res
      expect(data.createClient.client).toMatchSnapshot()
      expect(data.createClient.token.length).toBeGreaterThan(0)
      expect(errors).toBeUndefined()
    })
    it.skip('should deny creation of existing Client [Mutation]', async () => {
      const query = createClientQuery('TestDevie')
      const res = await request.anon(query)
      const { data, errors } = res
      expect(data).toBeNull()
      expect(errors.length).toBeGreaterThan(0)
    })
  })
  describe('Admin', async () => {
    let jwtToken = ''
    beforeAll(async () => {
      await seedDatabase(config.seeder)
      const expected = users[1]
      const query = {
        query: `mutation {
          login(data: {email: "${expected.email}", password: "password"}) {
            token
            user {
              id
              firstName
              lastName
              email
            }
          }
        }`,
      }
      const { data, errors } = await request.anon(query)
      data.login.token.should.be.a('string')
      expect(errors).toBeUndefined()
      const { login: { token } } = data
      jwtToken = token
    })
    it('should return all clients [Query]', async () => {
      const query = clientsQuery()
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
    it('should return client owned by User [Query]', async () => {
      const client = clients[3]
      const query = clientQuery(getSeedID(client))
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
    it('should return client not owned by User [Query]', async () => {
      const client = clients[1]
      const query = clientQuery(getSeedID(client))
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
    it('should create client [Mutation]', async () => {
      const query = createClientQuery('TestClient')
      const res = await request.anon(query)
      const { data, errors } = res
      expect(data.createClient.client).toMatchSnapshot()
      expect(data.createClient.token.length).toBeGreaterThan(0)
      expect(errors).toBeUndefined()
    })
    it('should update client owned by User [Mutation]', async () => {
      const client = clients[3]
      const domain = domains[0]
      const user = users[0]
      const query = updateClientQuery(getSeedID(client), 'RenamedTestClient', getSeedID(domain), [user.email])
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
    it('should update client not owned by User [Mutation]', async () => {
      const client = clients[1]
      const domain = domains[0]
      const user = users[0]
      const query = updateClientQuery(getSeedID(client), 'RenamedTestClient', getSeedID(domain), [user.email])
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
    it('should delete client owned by user [Mutation]', async () => {
      const client = clients[3]
      const query = deleteClientQuery(getSeedID(client))
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data.deleteClient.success).toBe(true)
    })
    it('should delete client not owned by user [Mutation]', async () => {
      const client = clients[1]
      const query = deleteClientQuery(getSeedID(client))
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data.deleteClient.success).toBe(true)
    })
  })
  describe('User', async () => {
    let jwtToken = ''
    beforeAll(async () => {
      await seedDatabase(config.seeder)
      const expected = users[0]
      const query = {
        query: `mutation {
          login(data: {email: "${expected.email}", password: "password"}) {
            token
            user {
              id
              firstName
              lastName
              email
            }
          }
        }`,
      }
      const { data, errors } = await request.anon(query)
      data.login.token.should.be.a('string')
      expect(errors).toBeUndefined()
      const { login: { token } } = data
      jwtToken = token
    })
    it('should return all clients [Query]', async () => {
      const query = clientsQuery()
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
    it('should return client owned by User [Query]', async () => {
      const client = clients[0]
      const query = clientQuery(getSeedID(client))
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
    it('should not return client not owned by User [Query]', async () => {
      const client = clients[3]
      const query = clientQuery(getSeedID(client))
      const { data, errors } = await request.user(query, jwtToken)
      expect(data).toBeNull()
      expect(errors.length).toBeGreaterThan(0)
    })
    it('should create client [Mutation]', async () => {
      const query = createClientQuery('TestClient')
      const res = await request.anon(query)
      const { data, errors } = res
      expect(data.createClient.client).toMatchSnapshot()
      expect(data.createClient.token.length).toBeGreaterThan(0)
      expect(errors).toBeUndefined()
    })
    it('should update client owned by User [Mutation]', async () => {
      const client = clients[0]
      const domain = domains[0]
      const user = users[0]
      const query = updateClientQuery(getSeedID(client), 'RenamedTestClient', getSeedID(domain), [user.email])
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
    it('should not update client not owned by User [Mutation]', async () => {
      const client = clients[3]
      const domain = domains[0]
      const user = users[0]
      const query = updateClientQuery(getSeedID(client), 'RenamedTestClient', getSeedID(domain), [user.email])
      const { data, errors } = await request.user(query, jwtToken)
      expect(data).toBeNull()
      expect(errors.length).toBeGreaterThan(0)
    })
    it('should delete client owned by user [Mutation]', async () => {
      const client = clients[0]
      const query = deleteClientQuery(getSeedID(client))
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data.deleteClient.success).toBe(true)
    })
    it('should not delete client not owned by user [Mutation]', async () => {
      const client = clients[3]
      const query = deleteClientQuery(getSeedID(client))
      const { data, errors } = await request.user(query, jwtToken)
      expect(data).toBeNull()
      expect(errors.length).toBeGreaterThan(0)
    })
  })
  describe('Client', async () => {
    let jwtToken = ''
    beforeAll(async () => {
      await seedDatabase(config.seeder)
      const query = {
        query: `mutation{createClient(data:{name:"TestClient"}){
          token
       }}`,
      }
      const { data, errors } = await request.anon(query)
      data.createClient.token.should.be.a('string')
      expect(errors).toBeUndefined()
      const { createClient: { token } } = data
      jwtToken = token
    })
    it('should not return all clients [Query]', async () => {
      const query = clientsQuery()
      const { data, errors } = await request.user(query, jwtToken)
      expect(data).toBeNull()
      expect(errors.length).toBeGreaterThan(0)
    })
    it('should not return client owned by User [Query]', async () => {
      const client = clients[0]
      const query = clientQuery(getSeedID(client))
      const { data, errors } = await request.user(query, jwtToken)
      expect(data).toBeNull()
      expect(errors.length).toBeGreaterThan(0)
    })
    it('should update client [Mutation]', async () => {
      const client = clients[0]
      const domain = domains[0]
      const user = users[0]
      const query = updateClientQuery(getSeedID(client), 'RenamedTestClient', getSeedID(domain), [getSeedID(user)])
      const { data, errors } = await request.user(query, jwtToken)
      expect(data).toBeNull()
      expect(errors.length).toBeGreaterThan(0)
    })
    it('should not delete client [Mutation]', async () => {
      const client = clients[3]
      const query = deleteClientQuery(getSeedID(client))
      const { data, errors } = await request.user(query, jwtToken)
      expect(data).toBeNull()
      expect(errors.length).toBeGreaterThan(0)
    })
  })
})
