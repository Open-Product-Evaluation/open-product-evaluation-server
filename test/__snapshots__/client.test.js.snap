// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client Admin should create client [Mutation] 1`] = `
Object {
  "domain": null,
  "name": "TestClient",
}
`;

exports[`Client Admin should return all clients [Query] 1`] = `
Object {
  "clients": Array [
    Object {
      "domain": Object {
        "id": "d8134f182c41d3cf4e9fdcda2be39855d7f1530757f06392c615fe652edee1a2",
      },
      "id": "5373ca535c2dbc81fbe1d239d84eeb27b9f695578e3268f9c208da8aa33cdeba",
      "name": "Forum Fernseher",
      "owners": Array [
        Object {
          "id": "bb4028c0657dfd7092f316dcbf5b006dce87217141e1b974441f040126acc142",
        },
      ],
    },
    Object {
      "domain": Object {
        "id": "d8134f182c41d3cf4e9fdcda2be39855d7f1530757f06392c615fe652edee1a2",
      },
      "id": "25098855d11587e9e64775846a4d483cd4251503b11dc9b616ea37fb83511f1c",
      "name": "Forum Tablet",
      "owners": null,
    },
    Object {
      "domain": Object {
        "id": "624497235d090cf74bd2473155d6d65fb9f9395ec018192d181387e20f4cbf10",
      },
      "id": "ced700fbf3e7c6454b92e1dd80454973156070f780142b7e236f3cb91172aa3c",
      "name": "Mensa Kiosk",
      "owners": null,
    },
    Object {
      "domain": Object {
        "id": "624497235d090cf74bd2473155d6d65fb9f9395ec018192d181387e20f4cbf10",
      },
      "id": "92357cc7f1024deec8c76803b207232978f94595a18df2a0ff8e3653b392e9f7",
      "name": "Mensa Tablet",
      "owners": Array [
        Object {
          "id": "bb9b812794666578a86254afca49deef281fd5caf4566db9915cf4d5ecde53dd",
        },
      ],
    },
  ],
}
`;

exports[`Client Admin should return client not owned by User [Query] 1`] = `
Object {
  "client": Object {
    "domain": Object {
      "id": "d8134f182c41d3cf4e9fdcda2be39855d7f1530757f06392c615fe652edee1a2",
    },
    "id": "25098855d11587e9e64775846a4d483cd4251503b11dc9b616ea37fb83511f1c",
    "name": "Forum Tablet",
    "owners": null,
  },
}
`;

exports[`Client Admin should return client owned by User [Query] 1`] = `
Object {
  "client": Object {
    "domain": Object {
      "id": "624497235d090cf74bd2473155d6d65fb9f9395ec018192d181387e20f4cbf10",
    },
    "id": "92357cc7f1024deec8c76803b207232978f94595a18df2a0ff8e3653b392e9f7",
    "name": "Mensa Tablet",
    "owners": Array [
      Object {
        "id": "bb9b812794666578a86254afca49deef281fd5caf4566db9915cf4d5ecde53dd",
      },
    ],
  },
}
`;

exports[`Client Admin should set owner of client not owned by User [Mutation] 1`] = `
Object {
  "setClientOwner": Object {
    "client": Object {
      "owners": Array [
        Object {
          "id": "bb4028c0657dfd7092f316dcbf5b006dce87217141e1b974441f040126acc142",
        },
      ],
    },
  },
}
`;

exports[`Client Admin should set owner of client owned by User [Mutation] 1`] = `
Object {
  "setClientOwner": Object {
    "client": Object {
      "owners": Array [
        Object {
          "id": "bb9b812794666578a86254afca49deef281fd5caf4566db9915cf4d5ecde53dd",
        },
        Object {
          "id": "bb4028c0657dfd7092f316dcbf5b006dce87217141e1b974441f040126acc142",
        },
      ],
    },
  },
}
`;

exports[`Client Admin should update client not owned by User [Mutation] 1`] = `
Object {
  "updateClient": Object {
    "client": Object {
      "domain": Object {
        "id": "d8134f182c41d3cf4e9fdcda2be39855d7f1530757f06392c615fe652edee1a2",
      },
      "id": "25098855d11587e9e64775846a4d483cd4251503b11dc9b616ea37fb83511f1c",
      "name": "RenamedTestClient",
      "owners": null,
    },
  },
}
`;

exports[`Client Admin should update client owned by User [Mutation] 1`] = `
Object {
  "updateClient": Object {
    "client": Object {
      "domain": Object {
        "id": "d8134f182c41d3cf4e9fdcda2be39855d7f1530757f06392c615fe652edee1a2",
      },
      "id": "92357cc7f1024deec8c76803b207232978f94595a18df2a0ff8e3653b392e9f7",
      "name": "RenamedTestClient",
      "owners": Array [
        Object {
          "id": "bb9b812794666578a86254afca49deef281fd5caf4566db9915cf4d5ecde53dd",
        },
      ],
    },
  },
}
`;

exports[`Client Anonym should create a Client [Mutation] 1`] = `
Object {
  "domain": null,
  "name": "TestClient",
}
`;

exports[`Client User should create client [Mutation] 1`] = `
Object {
  "domain": null,
  "name": "TestClient",
}
`;

exports[`Client User should return all clients [Query] 1`] = `
Object {
  "clients": Array [
    Object {
      "domain": Object {
        "id": "d8134f182c41d3cf4e9fdcda2be39855d7f1530757f06392c615fe652edee1a2",
      },
      "id": "5373ca535c2dbc81fbe1d239d84eeb27b9f695578e3268f9c208da8aa33cdeba",
      "name": "Forum Fernseher",
      "owners": Array [
        Object {
          "id": "bb4028c0657dfd7092f316dcbf5b006dce87217141e1b974441f040126acc142",
        },
      ],
    },
  ],
}
`;

exports[`Client User should return client owned by User [Query] 1`] = `
Object {
  "client": Object {
    "domain": Object {
      "id": "d8134f182c41d3cf4e9fdcda2be39855d7f1530757f06392c615fe652edee1a2",
    },
    "id": "5373ca535c2dbc81fbe1d239d84eeb27b9f695578e3268f9c208da8aa33cdeba",
    "name": "Forum Fernseher",
    "owners": Array [
      Object {
        "id": "bb4028c0657dfd7092f316dcbf5b006dce87217141e1b974441f040126acc142",
      },
    ],
  },
}
`;

exports[`Client User should set owner of client owned by User [Mutation] 1`] = `
Object {
  "setClientOwner": Object {
    "client": Object {
      "owners": Array [
        Object {
          "id": "bb4028c0657dfd7092f316dcbf5b006dce87217141e1b974441f040126acc142",
        },
      ],
    },
  },
}
`;

exports[`Client User should update client owned by User [Mutation] 1`] = `
Object {
  "updateClient": Object {
    "client": Object {
      "domain": Object {
        "id": "d8134f182c41d3cf4e9fdcda2be39855d7f1530757f06392c615fe652edee1a2",
      },
      "id": "5373ca535c2dbc81fbe1d239d84eeb27b9f695578e3268f9c208da8aa33cdeba",
      "name": "RenamedTestClient",
      "owners": Array [
        Object {
          "id": "bb4028c0657dfd7092f316dcbf5b006dce87217141e1b974441f040126acc142",
        },
      ],
    },
  },
}
`;
